### PR TITLE
Garbage Collector Fixes/Tweaks/Cleanup/Documenting

### PR DESCRIPTION
**Pull request by Krausus**
First, let me take a moment to describe the current state of our garbage collection system:
* When an object is added to the GC queue, it is destroyed, but not flagged as destroyed.
* When the GC checks if it needs to hard delete an object, it sees that it hasn't been flagged as destroyed, and skips it.
  * The current GC **never hard deletes anything it queues.**
* When anything else checks whether something was destroyed, it never appears to be, because nothing is ever flagged as destroyed.
* Some things check if something is destroyed by checking one of **two** other GC-related variables, neither of which the current GC is even aware of.
  * Naturally, these things don't realize when something is destroyed, either.
  * Because these two other GC-related variables are unused by the garbage collector, they shouldn't even be in the code at all.
* Because of the previous two points, our major processing loops don't realize when something has been destroyed, and will continue trying to process it, causing runtimes, and also causing things to not soft delete (and remember - nothing is hard deleted by the GC!).
* Every time a movable atom is deleted, its deletion method is recorded into one of three lists.
  * These lists are otherwise totally unused, meaning it is impossible to look at the deletions being recorded in them.
    * **Every** deleted /atom/movable, recorded in a list **you can never view.**
* Datums can be sent into qdel, but they may not be queued, because the garbage collector refuses to queue datums.

It's in pretty bad shape. Which is why, after taking some inspiration from the garbage collectors of TG and VG, I've made some changes...
* Garbage collection code has been organized into separate files, in its own folder.
  * The controller datum definitions that absolutely don't belong in the garbage collector code were tossed in with the master controller code.
  * GC-related code and vars have been pulled in from other files.
* GC variables and code that we don't use have been removed.
  * Places where these variables were checked have been replaced with checks that actually work.
* The garbage collector will now actually flag things as destroyed.
* The garbage collector will now actually hard delete things.
* Various processing controllers will now actually detect things that have been destroyed, and stop processing them.
* The garbage collector can now queue datums.
  * Currently, no datums are coded to queue up - they're all deleted immediately.
* Del and qdel will now catch any runtimes that occur when trying to Destroy something, rather than choking on them.
* Deletion profiling now needs to be toggled on, and there's a verb to view the deletion tracking lists.
* Some basic documentation, about how garbage collection works and how to use the main procs, has been written.

Unfortunately, although this *will* make our garbage collector actually work properly, it doesn't fix the yet unknown (though probably large) number of things that aren't actually coded properly for garbage collection. We can worry about fixing those later, though.

### CODE REVIEW

**Comment by Earthdivine** [comment]
So no more 1k queued things for Hard delete because someone opens a PDA?

**Comment by Krausus** [comment]
I'm not sure about PDAs, but the queue did gradually grow over time as things were qdel'd. The important thing is, it'll actually empty out now, regardless of what adds to it.

File: code/modules/garbage collection/garbage_collector.dm [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+			break
+
+		var/datum/D = locate(refID)
+		// If the object still exists, and it's the same object, hard del it
+		if(D && D.gcDestroyed == destroyedAtTime)
+			if(remainingForceDelPerTick <= 0)
+				break
+
+			#ifdef GC_DEBUG
+			gcwarning("GC process force delete [D.type]")
```
Comment (tigercat2000) [inline code/modules/garbage collection/garbage_collector.dm]: Not sure if this works or not. Most #ifdef's use tabbing.

Comment (Krausus) [inline code/modules/garbage collection/garbage_collector.dm]: Preprocessor directives, like #ifdef, are handled before the code is actually compiled, so they have no effect on the indentation on their own.

**Pull request merged**